File: scase-core/src/main/scala/io/jobial/scase/core/MessageHandler.scala
package io.jobial.scase.core

trait MessageContext[F[_]] {
  def receiveResult[M](request: M): MessageReceiveResult[F, M]
}

trait MessageHandler[F[_], M] {

  type Handler = Function[M, F[Unit]]

  def handleMessage(implicit context: MessageContext[F]): Handler
}

object MessageHandler {
  
  def apply[F[_], M](handler: MessageContext[F] => Function[M, F[Unit]]) =
    new MessageHandler[F, M] {
      def handleMessage(implicit context: MessageContext[F]): Handler = handler(context)
    }
}

File: scase-core/src/main/java/io/jobial/scase/core/javadsl/MessageHandler.java
package io.jobial.scase.core.javadsl;

import cats.effect.IO;
import cats.effect.IO$;
import scala.Function1;

import static io.jobial.scase.core.javadsl.JavaUtils.javaFunctionToScala;
import static io.jobial.scase.core.javadsl.JavaUtils.javaRunnableToScala;

public interface MessageHandler<M> extends io.jobial.scase.core.MessageHandler<IO, M> {

    default Function1<M, IO> handleMessage(io.jobial.scase.core.MessageContext<IO> context) {
        return javaFunctionToScala(request -> IO$.MODULE$.delay(javaRunnableToScala(() -> handleMessage(request, new MessageContext(context)))));
    }

    void handleMessage(M request, MessageContext context);

}
